care/facility/models/patient_consultation.py

```diff
@@ -377,3 +388,103 @@ class ConsultationClinician(models.Model):
+    )
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["consultation", "type"],
+                name="unique_consultation_consent",
+                condition=models.Q(archived=False),
+            ),
+            models.CheckConstraint(
+                name="patient_code_status_required",
+                check=~models.Q(type=ConsentType.PATIENT_CODE_STATUS)
+                | models.Q(patient_code_status__isnull=False),
+            ),
+            models.CheckConstraint(
+                name="patient_code_status_not_required",
+                check=models.Q(type=ConsentType.PATIENT_CODE_STATUS)
+                | models.Q(patient_code_status__isnull=True),
+            ),
+        ]
+
+    def __str__(self) -> str:
+        return f"{self.consultation.patient.name} - {ConsentType(self.type).label}{' (Archived)' if self.archived else ''}"
+
+    def save(self, *args, **kwargs):
+        if self.archived:
+            files = FileUpload.objects.filter(
+                associating_id=self.external_id,
+                file_type=FileUpload.FileType.CONSENT_RECORD,
+                is_archived=False,
+            )
+            files.update(
+                is_archived=True,
+                archived_datetime=timezone.now(),
+                archive_reason="Consent Archived",
+                archived_by=self.archived_by,
+            )
+
+        super().save(*args, **kwargs)
+
+    @staticmethod
+    def has_write_permission(request):
+        return request.user.is_superuser or (
+            request.user.verified
+            and ConsultationRelatedPermissionMixin.has_write_permission(request)
+        )
+
+    def has_object_read_permission(self, request):
+        if not super().has_object_read_permission(request):
+            return False
+        return (
+            request.user.is_superuser
+            or (
+                self.consultation.patient.facility
+                and request.user in self.consultation.patient.facility.users.all()
+            )
+            or (
+                self.consultation.assigned_to == request.user
+                or request.user == self.consultation.patient.assigned_to
+            )
+            or (
+                request.user.user_type >= User.TYPE_VALUE_MAP["DistrictLabAdmin"]
+                and (
+                    self.consultation.patient.facility
+                    and request.user.district
+                    == self.consultation.patient.facility.district
+                )
+            )
+            or (
+                request.user.user_type >= User.TYPE_VALUE_MAP["StateLabAdmin"]
+                and (
+                    self.consultation.patient.facility
+                    and request.user.state == self.consultation.patient.facility.state
+                )
+            )
+        )
+
+    def has_object_update_permission(self, request):
+        return super().has_object_update_permission(
```

Comment:
the default update permission only checks if the user is >= nurse, the current conditions will allow non-home facility users to update the records

Reply:
but wouldn't `self.has_object_read_permission(request)` look out for that?